### DISCUSSION TITLE
Is the floating point stack optional or not?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ForthHub/discussion#130] [Q&A] [answered]
Both the Forth 2012 and the Draft state in section 12.3.3 Floating-point stack:

> the floating-point stack `shall` be separate from the data and return stacks.

Then in 12.3.4 Environmental queries:

> FLOATING-STACK n yes the maximum depth of the separate floating-point stack. On systems with the environ-mental restriction of `keeping floating-point items on the data stack, n = 0.`

Which allows for the option of using the data-stack instead.  So what is the actual intent:  more restrictive MUST use a float stack or MAY use the data-stack for floats for more flexible implementation choice?  Seems like the next Draft should clarify this, maybe with Rationale background.

**Comment by ruv** [comment, marked as answer]
The Forth standard implement conceptions of environmental restrictions / environmental dependencies.
  - A program that complies with some environmental restriction will work identically on systems with and without this restriction.
  - A program without some environmental dependency will work identically regardless whether a system fulfill this dependency or does not fulfill.

>  So what is the actual intent: more restrictive MUST use a float stack or MAY use the data-stack for floats for more flexible implementation choice? 

Have a look at [C.7.2 Separate Floating-point Stack is now Standard](https://forth-standard.org/standard/diff#diff:fpstack):

  - > Forth 94 systems that implement a combined stack become systems with an environmental restriction of not providing a separate floating-point stack, but a combined stack. 

A system may provide the floating-point stack combined with the data stack, but then this system cannot be a **Forth-2012 System**, but only a **Forth-2012 System with Environmental Restrictions**.

See also [5 Compliance and labeling](https://forth-standard.org/standard/label), [12.4.1.4 Environmental restrictions](https://forth-standard.org/standard/float#float:restrictions) (in The optional Floating-Point word set).

**Reply by SirWumpus** [reply]
So Forth 2012 and the Draft prefer migrating to separate float and data stacks.  I suppose this is because while modern 32b or 64b consumer CPUs can handle floats in two or one cell, embedded systems may still rely on smaller cell sizes that make a shared float/data stack difficult for programs to deal with if I understood the referenced sections above.

**Reply by ruv** [reply]
I don't see anything concerning smaller cell size. It says (emphasized by me):
- > **Reason:**  The developing of software that may be used with either a combined stack or a separate stack is **extremely difficult and costly**.

The reason of this difficulty is that an ANS Forth Program should work on an ANS Forth System regardless whether the system provides a separate floating-point stack, or combined.

**Reply by SirWumpus** [reply]
I'm confused.  Are you trying to say that it does not matter which float stack implementation (separate or combined) is used by a system provided that standard user programs can not tell (_nor rely on_) the difference -- the system should _appear_ to have a separate float stack from the data stack from a standard user program's view point?

**Reply by ruv** [reply]
"ANS Forth" — is an [official name](http://lars.nocrew.org/dpans/dpans5.htm) for Forth-94, they are synonyms now.

A Forth-94 System, if it provides the optional Floating-Point word set, is allowed to provide either a combined floating-point stack, or separate.

A Forth-2012 System, if it provides the optional Floating-Point word set, should provide a separate floating-point stack (or, otherwise, it can be not more than a Forth-2012 System with Environmental Restrictions).

> the system should _appear_ to have a separate float stack from the data stack from a standard user program's view point?

It seems to be true.
